syntax = "proto3";

package keyserver;

import "google/protobuf/empty.proto";

// Key Server Service
service KeyServer {
  rpc ReleaseKey(ReleaseKeyRequest) returns (google.protobuf.Empty);
  rpc ReloadKeys(ReloadKeysRequest) returns (ReloadKeysResponse);
  rpc GetStats(google.protobuf.Empty) returns (StatsResponse);
  rpc GetDetailedStats(google.protobuf.Empty) returns (DetailedStatsResponse);
  rpc TrackUsage(TrackUsageRequest) returns (TrackUsageResponse);
  rpc Health(google.protobuf.Empty) returns (HealthResponse);
  rpc SelectKeysForSession(SelectKeysRequest) returns (SelectKeysResponse);
}

message ReleaseKeyRequest {
  string session_id = 1;
}

message ReloadKeysRequest {
  string file_path = 1;
}

message ReloadKeysResponse {
  bool success = 1;
  map<string, int32> pools = 2;
  string error = 3;
}

// Updated for string format
message SelectKeysRequest {
  string session_id = 1;
  int32 user_id = 2;
  repeated string models = 3;  // Now simple strings: "provider/model"
  int32 count_per_model = 4;
}

message SelectedKey {
  string key_id = 1;
  string provider = 2;
  string model = 3;
  string api_key = 4;
  int32 tokens_hour = 6;
  int32 tokens_total = 7;
  string status = 8;
}

message SelectKeysResponse {
  bool success = 1;
  repeated SelectedKey keys = 2;
  string error = 3;
}

message StatsResponse {
  bool success = 1;
  map<string, PoolStats> pool_stats = 2;
  RuntimeStats runtime_stats = 3;
  string error = 4;
}

message DetailedStatsResponse {
  bool success = 1;
  map<string, PoolKeyStats> pool_detailed_stats = 2;
  RuntimeStats runtime_stats = 3;
  string error = 4;
}

message PoolStats {
  int32 available = 1;
}

message PoolKeyStats {
  repeated KeyStats keys = 1;
}

message KeyStats {
  string key_id = 1;
  int32 tokens_hour = 2;
  int32 tokens_total = 3;
  int64 last_used = 4;
}

message RuntimeStats {
  int64 total_requests = 1;
  int64 successful_requests = 2;
  int64 failed_requests = 3;
  double uptime_seconds = 4;
}

message HealthResponse {
  bool success = 1;
  bool healthy = 2;
}

message TrackUsageRequest {
  string key_id = 1;
  int32 tokens_used = 2;
}

message TrackUsageResponse {
  bool success = 1;
  string error = 2;
} 